src/ILCompiler.Compiler/src/IL/Stubs/StartupCode/StartupCodeMainMethod.cs

```diff
@@ -0,0 +1,107 @@
+﻿using Internal.TypeSystem;
+using Internal.IL;
+using Internal.IL.Stubs;
+using System;
+
+namespace Internal.IL.Stubs.StartupCode
+{
+    /// <summary>
+    /// Startup code that does initialization, Main invocation
+    /// and shutdown of the runtime.
+    /// </summary>
+    public sealed class StartupCodeMainMethod : ILStubMethod
+    {
+        private TypeDesc _owningType;
+        private MethodDesc _mainMethod;
+        private MethodSignature _signature;
+
+        public StartupCodeMainMethod(TypeDesc owningType, MethodDesc mainMethod)
+        {
+            _owningType = owningType;
+            _mainMethod = mainMethod;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return OwningType.Context;
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get
+            {
+                return _owningType;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return "StartupCodeMain";
+            }
+        }
+
+        public override MethodIL EmitIL()
+        {
+            ILEmitter emitter = new ILEmitter();
+            ILCodeStream codeStream = emitter.NewCodeStream();
+            ILLocalVariable returnValue = emitter.NewLocal(Context.GetWellKnownType(WellKnownType.Int32));
+
+            TypeDesc startup = Context.SystemModule.GetType("Internal.Runtime.CompilerHelpers", "StartupCodeHelpers");
+            
+            codeStream.Emit(ILOpcode.call, emitter.NewToken(startup.GetMethod("Initialize", null)));
+
+            // Initialize command line args
+            // TODO: For Windows change to "InitializeCommandLineArgsW" with wmain wchar_t change.
+            string initArgsName = (Context.Target.OperatingSystem == TargetOS.Windows)
+                                ? "InitializeCommandLineArgs"
+                                : "InitializeCommandLineArgs";
+            MethodDesc initArgs = startup.GetMethod(initArgsName, null);
+            codeStream.Emit(ILOpcode.ldarg_0); // argc
+            codeStream.Emit(ILOpcode.ldarg_1); // argv
+            codeStream.Emit(ILOpcode.call, emitter.NewToken(initArgs));
+
+            // Call program Main
+            if (_mainMethod.Signature.Length > 0)
+            {
+                TypeDesc environ = Context.SystemModule.GetType("System", "Environment");
+                codeStream.Emit(ILOpcode.call, emitter.NewToken(environ.GetMethod("GetCommandLineArgs", null)));
+            }
+            codeStream.Emit(ILOpcode.call, emitter.NewToken(_mainMethod));
+            if (_mainMethod.Signature.ReturnType.IsVoid)
+            {
+                codeStream.EmitLdc(0);
+            }
+            codeStream.EmitStLoc(returnValue);
+
+            codeStream.Emit(ILOpcode.call, emitter.NewToken(startup.GetMethod("Shutdown", null)));
```

Comment:
For an exception gone unhandled, either the OS implementation will result in an unwind or not. If it does not, Shutdown will never get invoked. But if it does, then the question is if we expect Shutdown to execute in such a case?

I think that is a fine place to be in, where we do not invoke Shutdown for such scenarios, provided that is a semantic we want to drive (am equating Shutdown here to EEShutdown that happens in CLR when DllProcessDetach happens even for unhandled exceptions).